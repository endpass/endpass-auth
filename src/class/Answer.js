export default class Answer {
  static createOk(payload) {
    return {
      status: true,
      payload,
    };
  }

  // TODO: `error` field support < 0.24 version
  static createFail(code, error) {
    return {
      status: false,
      code,
      error,
    };
  }
}
