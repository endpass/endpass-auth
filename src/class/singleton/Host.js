import { parseUrl } from '@/util/dom';

const DEFAULT_ORIGIN = parseUrl(document.referrer).origin;

class Host {
  constructor() {
    this.value = DEFAULT_ORIGIN;
  }

  set origin(value) {
    if (Object.isFrozen(this)) return;

    const defineValue = value || DEFAULT_ORIGIN;
    this.value = parseUrl(defineValue).origin;

    Object.freeze(this);
  }

  get origin() {
    return this.value;
  }
}

export { Host };

export default new Host();
