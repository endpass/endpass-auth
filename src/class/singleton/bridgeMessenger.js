import CrossWindowMessenger from '@endpass/class/CrossWindowMessenger';
import { DIRECTION } from '@/constants';

const bridgeMessenger = new CrossWindowMessenger({
  // showLogs: !ENV.VUE_APP_IS_PRODUCTION,
  name: 'auth-bridge',
  target: window.parent,
  from: DIRECTION.AUTH,
  to: DIRECTION.CONNECT,
});

export default bridgeMessenger;
