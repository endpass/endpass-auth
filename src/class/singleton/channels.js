import Channel from '@/class/Channel';

export const authChannel = new Channel('auth');
export const signChannel = new Channel('sign');
export const accountChannel = new Channel('account');
export const permissionChannel = new Channel('permission');
export const documentChannel = new Channel('document');
export const walletChannel = new Channel('wallet');

export default {
  authChannel,
  signChannel,
  accountChannel,
  permissionChannel,
  documentChannel,
  walletChannel,
};
