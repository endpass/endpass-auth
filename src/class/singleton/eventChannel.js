import EventChannel from '@/class/EventChannel';

export default new EventChannel();
