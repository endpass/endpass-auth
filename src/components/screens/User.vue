<template>
  <screen @close="handleWindowClose">
    <v-modal-card
      :loading="!isInited"
      :closable="isDialog"
      @close="handleCancel"
    >
      <account-form
        :loading="loading"
        :closable="isDialog"
        :accounts="accountsOptions"
        :networks="networksOptions"
        :form-data="formData"
        :can-logout="!isDemoMode"
        :error="error"
        :message="message"
        @donate-request="handleDonateRequest"
        @donate-success="handleDonateSuccess"
        @donate-error="handleDonateError"
        @submit="handleAccountFormSubmit"
        @cancel="handleCancel"
        @logout="handleLogout"
      />
    </v-modal-card>
  </screen>
</template>

<script>
import { mapActions, mapMutations, mapState, mapGetters } from 'vuex';
import Network from '@endpass/class/Network';
import Screen from '@/components/common/Screen';
import AccountForm from '@/components/forms/Account';
import VModalCard from '@endpass/ui/kit/VModalCard';

export default {
  name: 'User',

  data: () => ({
    formData: {
      activeAccount: null,
      activeNet: null,
    },
    error: null,
    message: null,
  }),

  computed: {
    ...mapState({
      isInited: state => state.core.isInited,
      loading: state => state.core.loading,
      settings: state => state.accounts.settings,
      isDemoMode: state => !!state.accounts.demoData,
      accounts: state => state.accounts.accounts,
    }),
    ...mapGetters(['isDialog']),

    networksOptions() {
      return Object.values(Network.DEFAULT_NETWORKS).map(({ id, name }) => ({
        val: id,
        text: name,
      }));
    },

    accountsOptions() {
      if (!this.accounts) return [];

      return this.accounts.map(({ address }) => ({
        val: address,
        text: address,
      }));
    },
  },

  watch: {
    formData: {
      handler() {
        if (this.message) {
          this.message = null;
        }

        if (this.error) {
          this.error = null;
        }
      },
      deep: true,
    },
  },

  methods: {
    ...mapMutations(['changeLoadingStatus']),
    ...mapActions(['logout', 'closeAccount', 'updateSettings', 'dialogClose']),

    async handleAccountFormSubmit() {
      const { activeAccount, activeNet } = this.formData;

      try {
        this.error = null;

        await this.updateSettings({
          lastActiveAccount: activeAccount,
          net: activeNet,
        });
      } catch (err) {
        console.error(err);

        this.error = err.message;
      }
    },

    async handleLogout() {
      this.logout();
    },

    handleCancel() {
      this.closeAccount();
      this.dialogClose();
    },

    handleDonateRequest() {
      this.changeLoadingStatus(true);
    },

    handleDonateSuccess() {
      this.message = this.$i18n.t('components.user.donationSuccess');
      this.changeLoadingStatus(false);
    },

    handleDonateError(e) {
      this.error = e;
      this.changeLoadingStatus(false);
    },

    handleWindowClose() {
      this.closeAccount();
      this.dialogClose();
    },
  },

  async created() {
    const { settings } = this;
    this.formData.activeAccount = settings.lastActiveAccount;
    this.formData.activeNet = settings.net;
  },

  components: {
    VModalCard,
    Screen,
    AccountForm,
  },
};
</script>
