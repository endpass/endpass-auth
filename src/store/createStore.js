// @ts-check
import Vue from 'vue';
import Vuex from 'vuex';

Vue.use(Vuex);

/**
 * @param {object?} [options] passed options for store
 * @return {import('vuex').Store<{}>} store Vuex store instance
 */
export default options => {
  return new Vuex.Store({
    strict: !ENV.VUE_APP_IS_PRODUCTION,
    ...options,
  });
}
