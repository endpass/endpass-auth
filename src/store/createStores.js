import GasPriceModule from '@/store/modules/GasPriceModule';
import SharedModule from '@/store/modules/SharedModule';
import AccountsModule from '@/store/modules/AccountsModule';
import CoreModule from '@/store/modules/CoreModule';
import RequestsModule from '@/store/modules/RequestsModule';
import createStoreModule from '@/store/createStoreModule';

export default store => {
  const createModule = createStoreModule(store);

  const gasPriceStore = createModule(GasPriceModule, 'gasPrice');

  const sharedStore = createModule(SharedModule, 'shared');

  const accountsStore = createModule(AccountsModule, 'accounts', {
    sharedStore,
  });

  const coreStore = createModule(CoreModule, 'core', {
    accountsStore,
    sharedStore,
  });

  const requestStore = createModule(RequestsModule, 'requests', {
    accountsStore,
    sharedStore,
  });

  return {
    gasPriceStore,
    sharedStore,
    accountsStore,
    coreStore,
    requestStore,
  }
}
