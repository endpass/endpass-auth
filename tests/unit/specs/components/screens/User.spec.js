import Vuex from 'vuex';
import { shallowMount, createLocalVue } from '@vue/test-utils';
import User from '@/components/screens/User';
import setupI18n from '@/locales/i18nSetup';
import { accountChannel } from '@/class/singleton/channels';
import Answer from '@/class/Answer';
import bridgeMessenger from '@/class/singleton/bridgeMessenger';
import createStore from '@/store/createStore';
import createStores from '@/store/createStores';

const localVue = createLocalVue();

localVue.use(Vuex);
const i18n = setupI18n(localVue);

describe('User', () => {
  let wrapper;
  let accountsStore;

  beforeEach(async () => {
    bridgeMessenger.sendAndWaitResponse.mockResolvedValueOnce({});

    const store = createStore();
    const {
      accountsStore: accountStoreModule,
      coreStore,
      sharedStore,
    } = createStores(store);
    accountsStore = accountStoreModule;

    await accountsStore.defineSettings();
    await accountsStore.defineOnlyV3Accounts();

    wrapper = shallowMount(User, {
      accountsStore,
      coreStore,
      sharedStore,
      localVue,
      i18n,
    });
  });

  describe('render', () => {
    it('should correctly render User component', () => {
      expect(wrapper.name()).toBe('User');
      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  describe('behavior', () => {
    it('should set form data from settings on create', () => {
      expect(wrapper.vm.formData).toEqual({
        activeAccount: '0x0',
        activeNet: 1,
      });
    });

    it('should update settings of form submit', async () => {
      expect.assertions(1);

      const dataPromise = accountChannel.take();
      const { activeAccount, activeNet } = wrapper.vm.formData;

      wrapper.find('account-form-stub').vm.$emit('submit');

      await global.flushPromises();
      const res = await dataPromise;

      expect(res).toEqual(
        Answer.createOk({
          settings: {
            activeAccount,
            activeNet,
          },
          type: 'update',
        }),
      );
    });

    it('should logout if logout button was pressed in form', async () => {
      expect.assertions(2);

      expect(accountsStore.isLogin).toBe(true);

      wrapper.find('account-form-stub').vm.$emit('logout');

      await global.flushPromises();

      expect(accountsStore.isLogin).toBe(false);
    });

    it('should close account if cancel button was pressed in form', async () => {
      expect.assertions(1);

      const dataPromise = accountChannel.take();

      wrapper.find('account-form-stub').vm.$emit('cancel');
      const res = await dataPromise;
      expect(res).toEqual(Answer.createOk({ type: 'close' }));
    });
  });
});
