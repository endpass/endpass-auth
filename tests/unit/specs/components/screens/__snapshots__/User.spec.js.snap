// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User render should correctly render User component 1`] = `
<screen-stub>
  <v-modal-card-stub isclosable="true">
    <account-form-stub isclosable="true" accounts="[object Object]" networks="[object Object],[object Object],[object Object],[object Object]" formdata="[object Object]"></account-form-stub>
  </v-modal-card-stub>
</screen-stub>
`;
